src/client/api/object.c

```diff
@@ -440,3 +440,368 @@ daos_oit_list(daos_handle_t oh, daos_obj_id_t *oids, uint32_t *oids_nr,
+				*oids_nr = *oids_nr - 1;
+				i--;
+				continue;
+			} else {
+				D_ASSERTF(rc == 1, "bad rc %d\n", rc);
+				rc = 0;
+				continue;
+			}
+		}
+		oit_filter_arg_free(oa);
+		return rc;
+	}
+
+	rc = oit_filter_fetch_init(oa);
+	if (rc) {
+		oit_filter_arg_free(oa);
+		return rc;
+	}
+
+	/* create fetch task to get back marker data, cannot directly call dc_obj_fetch_task_create
+	 * here to avoid dc_task_create() -> daos_event_priv_get() failure.
+	 */
+	DAOS_API_ARG_ASSERT(*fargs, OBJ_FETCH);
+	rc = tse_task_create(dc_obj_fetch_task, tse_task2sched(task), NULL, &ftask);
+	if (rc) {
+		oit_filter_arg_free(oa);
+		return rc;
+	}
+	task_args = tse_task_buf_embedded(ftask, sizeof(struct daos_task_args));
+	task_args->ta_magic = DAOS_TASK_MAGIC;
+	fargs = (void *)&task_args->ta_u;
+	fargs->oh		= oa->oa_oh;
+	fargs->th		= DAOS_TX_NONE;
+	fargs->flags		= 0;
+	fargs->dkey		= &oa->oa_dkey;
+	fargs->nr		= *oids_nr;
+	fargs->extra_flags	= 0;
+	fargs->iods		= oa->oa_fiods;
+	fargs->sgls		= oa->oa_fsgls;
+	fargs->ioms		= NULL;
+	fargs->extra_arg	= NULL;
+	fargs->csum_iov		= NULL;
+
+	rc = tse_task_register_deps(task, 1, &ftask);
+	if (rc != 0) {
+		tse_task_complete(ftask, rc);
+		oit_filter_arg_free(oa);
+		return rc;
+	}
+
+	rc = tse_task_register_comp_cb(task, oit_filter_cb, &oa, sizeof(oa));
+	if (rc) {
+		tse_task_complete(ftask, rc);
+		oit_filter_arg_free(oa);
+		return rc;
+	}
+
+	oa->oa_fetched = 1;
+	rc = tse_task_schedule(ftask, 0);
+	if (rc)
+		tse_task_complete(ftask, rc);
+
+	return 0;
+}
+
+int
+daos_oit_list_filter(daos_handle_t oh, daos_obj_id_t *oids, uint32_t *oids_nr,
+		     daos_anchor_t *anchor, daos_oit_filter_cb filter, daos_event_t *ev)
+{
+	struct oit_filter_arg	*oa;
+	tse_task_t		*task;
+	int			 i;
+	int			 rc;
+
+	if (daos_handle_is_inval(oh) || oids == NULL || oids_nr == NULL || *oids_nr == 0 ||
+	    anchor == NULL || filter == NULL)
+		return -DER_INVAL;
+
+	D_ALLOC_PTR(oa);
+	if (!oa)
+		return -DER_NOMEM;
+
+	oa->oa_oh	= oh;
+	oa->oa_oids	= oids;
+	oa->oa_oids_nr	= oids_nr;
+	oa->oa_ev	= ev;
+	oa->oa_filter	= filter;
+	oa->oa_nr	= *oids_nr;
+	oa->oa_fetched	= false;
+	D_ALLOC_ARRAY(oa->oa_kds, oa->oa_nr);
+	if (!oa->oa_kds)
+		D_GOTO(failed, rc = -DER_NOMEM);
+
+	rc = d_sgl_init(&oa->oa_sgl, oa->oa_nr);
+	if (rc)
+		D_GOTO(failed, rc = -DER_NOMEM);
+
+	for (i = 0; i < oa->oa_nr; i++)
+		d_iov_set(&oa->oa_sgl.sg_iovs[i], &oids[i], sizeof(oids[i]));
+
+	DAOS_OIT_DKEY_SET(&oa->oa_dkey, &oa->oa_bucket);
+	rc = dc_obj_list_akey_task_create(oh, DAOS_TX_NONE, &oa->oa_dkey,
+					  oids_nr, oa->oa_kds, &oa->oa_sgl,
+					  anchor, ev, NULL, &task);
+	if (rc)
+		D_GOTO(failed, rc);
+
+	rc = tse_task_register_comp_cb(task, oit_filter_cb, &oa, sizeof(oa));
+	if (rc) {
+		tse_task_complete(task, rc);
+		D_GOTO(failed, rc);
+	}
+
+	return dc_task_schedule(task, true);
+
+failed:
+	/* NB: OK to call with empty sgl */
+	oit_filter_arg_free(oa);
+	return rc;
+}
+
+int
+oit_filter_unmarked(daos_obj_id_t oid, d_iov_t *marker)
```

Comment:
adding filter to server-side is considerable later, but that will cause a problem that the list_unmark cannot work with old server, and also cause RPC inter-operability problem. So let's provide a basic implementation without that two problems now.

Reply:
sure to consider later and not block the PR, but why would there be any RPC incompatibility? we can just add a new RPC for list with filter. and even if there is any incompatibility, it would have to be done sooner or later.

Reply:
I was though to reuse existing RPC rather than adding new RPC.